### PR TITLE
access host dir of /proc when running kubelet as a container

### PR DESCRIPTION
**Pull request by dalanlan**
Fix syncing pod issue of containerized kubelet. Part of #13627 
This uses a simple trick as cadvisor --- 
to judge whether kubelet is running in a container based on the mounted volume. (locates https://github.com/google/cadvisor/blob/master/manager/manager.go#L135~L140), which based on the assumption that we have the host's root filesystem available at /rootfs ( specified at https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/nsenter_mount.go#L42)

Now we can get rid of following syncing pod problem

```
$ kubectl describe pods/busybox
...
Events:
  FirstSeen             LastSeen            Count   From            SubobjectPath               Reason      Message
  Sun, 06 Sep 2015 17:12:55 +0800   Sun, 06 Sep 2015 17:12:55 +0800 1   {scheduler }                            Scheduled   Successfully assigned busybox to 10.10.102.28
  Sun, 06 Sep 2015 17:12:55 +0800   Sun, 06 Sep 2015 17:12:55 +0800 1   {kubelet 10.10.102.28}  implicitly required container POD   pulled      Container image "gcr.io/google_containers/pause:0.8.0" already present on machine
  Sun, 06 Sep 2015 17:12:55 +0800   Sun, 06 Sep 2015 17:12:55 +0800 1   {kubelet 10.10.102.28}  implicitly required container POD   Created     Created with docker id fa5db056dd41
  Sun, 06 Sep 2015 17:12:55 +0800   Sun, 06 Sep 2015 17:12:55 +0800 1   {kubelet 10.10.102.28}  implicitly required container POD   Started     Started with docker id fa5db056dd41
  Sun, 06 Sep 2015 17:12:55 +0800   Sun, 06 Sep 2015 17:12:55 +0800 1   {kubelet 10.10.102.28}                      FailedSync  Error syncing pod, skipping: open /proc/27373/cgroup: no such file or directory
```

Yet it _doesn't_ address everything, we now have this

```
$ kubectl describe pods/busybox
...
Events:
  FirstSeen             LastSeen            Count   From            SubobjectPath               Reason      Message
  Tue, 08 Sep 2015 17:35:49 +0800   Tue, 08 Sep 2015 17:35:49 +0800 1   {scheduler }                            Scheduled   Successfully assigned busybox1 to 10.10.102.28
  Tue, 08 Sep 2015 17:35:49 +0800   Tue, 08 Sep 2015 17:35:49 +0800 1   {kubelet 10.10.102.28}  implicitly required container POD   pulled      Container image "gcr.io/google_containers/pause:0.8.0" already present on machine
  Tue, 08 Sep 2015 17:35:49 +0800   Tue, 08 Sep 2015 17:35:49 +0800 1   {kubelet 10.10.102.28}  implicitly required container POD   Created     Created with docker id 9e50e5ebf978
  Tue, 08 Sep 2015 17:35:49 +0800   Tue, 08 Sep 2015 17:35:49 +0800 1   {kubelet 10.10.102.28}  implicitly required container POD   Started     Started with docker id 9e50e5ebf978
  Tue, 08 Sep 2015 17:35:49 +0800   Tue, 08 Sep 2015 17:35:49 +0800 1   {kubelet 10.10.102.28}                      FailedSync  Error syncing pod, skipping: exceeded maxTries, some processes might not have desired OOM score
  Tue, 08 Sep 2015 17:35:59 +0800   Tue, 08 Sep 2015 17:35:59 +0800 1   {kubelet 10.10.102.28}  spec.containers{busybox}        pulled      Container image "busybox" already present on machine
  Tue, 08 Sep 2015 17:35:59 +0800   Tue, 08 Sep 2015 17:35:59 +0800 1   {kubelet 10.10.102.28}  spec.containers{busybox}        Created     Created with docker id 8e1ce01ce59c
  Tue, 08 Sep 2015 17:35:59 +0800   Tue, 08 Sep 2015 17:35:59 +0800 1   {kubelet 10.10.102.28}  spec.containers{busybox}        Started     Started with docker id 8e1ce01ce
```

And i would love to dive deep into oom later.
/cc @dchen1107

### CODE REVIEW

File: pkg/util/procfs/procfs.go [hunk 1]
```diff
@@ -45,7 +46,15 @@ func containerNameFromProcCgroup(content string) (string, error) {
 // Eg. If the devices cgroup for the container is stored in /sys/fs/cgroup/devices/docker/nginx,
 // return docker/nginx. Assumes that the process is part of exactly one cgroup hierarchy.
 func (pfs *ProcFs) GetFullContainerName(pid int) (string, error) {
-	filePath := path.Join("/proc", strconv.Itoa(pid), "cgroup")
+	var filePath string
+	// if kubelet was started with host's rootfs mounted, assume it's running as a container
+	_, err := os.Stat("/rootfs/proc")
```
Comment (vishh) [inline pkg/util/procfs/procfs.go]: Is `/rootfs` defined as the required mountpoint somewhere?

File: pkg/util/procfs/procfs.go [hunk 2]
```diff
@@ -45,7 +46,15 @@ func containerNameFromProcCgroup(content string) (string, error) {
 // Eg. If the devices cgroup for the container is stored in /sys/fs/cgroup/devices/docker/nginx,
 // return docker/nginx. Assumes that the process is part of exactly one cgroup hierarchy.
 func (pfs *ProcFs) GetFullContainerName(pid int) (string, error) {
-	filePath := path.Join("/proc", strconv.Itoa(pid), "cgroup")
+	var filePath string
+	// if kubelet was started with host's rootfs mounted, assume it's running as a container
```
Comment (vishh) [inline pkg/util/procfs/procfs.go]: Even if kubelet is running in a container, AFAIK it currently assumes it has control over the entire node. I hope those assumptions are not an issue for your use case.

**Comment by ncdc** [comment]
@kubernetes/rh-cluster-infra

File: pkg/util/procfs/procfs.go (hunk 1 above)
Comment (pmorie) [inline pkg/util/procfs/procfs.go]: @vishh I think it is documented only in `nsenter_mount.go` and `local-up-cluster.sh` -- we probably need to have a guidance doc for running the kubelet in a container that summarizes the current state and assumptions.

Comment (pmorie) [inline pkg/util/procfs/procfs.go]: I wonder if we should inject the `containerized` option from the kubelet in this context and use that to drive the path construction

Comment (vishh) [inline pkg/util/procfs/procfs.go]: @pmorie This path could be made configurable too. WDYT? 
+1 for injecting a `containerized` option.

Comment (pmorie) [inline pkg/util/procfs/procfs.go]: @vishh Maybe we should just inject the path and drive it from outside based
on whether kubelet is containerized, that way the 'containerized' logic
doesn't creep into ProcPs

Comment (dchen1107) [inline pkg/util/procfs/procfs.go]: +1 on what Paul proposed above.

Comment (dalanlan) [inline pkg/util/procfs/procfs.go]: Fair enough.

**Comment by pmorie** [comment]
@dalanlan ping?

**Comment by dalanlan** [comment]
@pmorie Sorry but i was stuck by other business recently.
I would love to get into this, yet i do want to confirm one thing. Should we make the mount path (/rootfs, hard coded for now) configurable, or just inject the optional path("/rootfs/proc" for containerized kubelet and "/proc" otherwise)?

**Comment by fgrzadkowski** [comment]
@dalanlan Isn't it that it's enough to pass ```--pid=host``` option when starting kubelet? See https://docs.docker.com/reference/commandline/run/

**Comment by vishh** [comment]
@fgrzadkowski: Some parts of kubelet access procfs directly. Other parts using linux utilities like `pidof`. I'd argue that we want all of kubelet to access process information via procfs. As of now, both this PR and passing `--pid=host` is necessary, which is unfortunate.

**Comment by fgrzadkowski** [comment]
@vishh @dalanlan If you pass ```--pid=host``` then both ```procfs``` and ```pidof``` will work from within container. Please see my example:
```
$ docker ps
CONTAINER ID        IMAGE                       COMMAND             CREATED             STATUS              PORTS               NAMES
118322c5d863        kubernetes/serve_hostname   "/serve_hostname"   3 minutes ago       Up 3 minutes        9376/tcp            loving_mclean       
2ce1880517f4        kubernetes/serve_hostname   "/serve_hostname"   3 minutes ago       Up 3 minutes        9376/tcp            thirsty_goodall     
$ ps aux | grep serve_hostname | grep -v grep
root     12921  0.0  0.0   7152  2064 ?        Ssl  21:05   0:00 /serve_hostname
root     12963  0.0  0.0   7152  2060 ?        Ssl  21:05   0:00 /serve_hostname
$ docker run -it busybox /bin/sh -c "ps aux | grep serve_hostname | grep -v grep" 
$ docker run -it --pid=host busybox /bin/sh -c "ps aux | grep serve_hostname | grep -v grep" 
12921 root     /serve_hostname
12963 root     /serve_hostname
$ docker run -it --pid=host busybox /bin/sh -c "pidof serve_hostname" 
12963 12921
```

**Comment by vishh** [comment]
Hmm. If running kubelet in the host pid namespace is ok in general, we can switch to not relying on host procfs being mounted.

**Comment by pmorie** [comment]
@vishh I think we will still need proc being mounted to read the host mount ns from

**Comment by vishh** [comment]
@pmorie: Are you referring to `/proc/<pid>/root/`?

**Comment by pmorie** [comment]
@vishh `/proc/1/ns`

**Comment by vishh** [comment]
Hmm. With `--pid=host` and `--privileged` set, we should be able to access
everything from the host procfs right?

**Comment by dalanlan** [comment]
Yes indeed. Good findings:) Closing this.
Yet it still has follow-up problems of oom, which can be discussed in #14402

**Pull request closed without merging**